### PR TITLE
#250 - Support for caching Docker proxy

### LINKED ISSUE #250: Caching Docker proxy
Implement caching proxy Docker registry https://github.com/artipie/docker-adapter/issues/172 and include it to Artipie next release.

### PR DESCRIPTION
**Pull request by olegmoz**
Closes #250 
Added support for caching to `docker-proxy` repo type

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#281](https://github.com/artipie/artipie/issues/281) is assigned to @vatavuk[/z](https://www.0crat.com/u/vatavuk), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/3d68328c-a330-45e4-be37-7426285f823a); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: src/main/java/com/artipie/SliceFromConfig.java [hunk 1]
```diff
@@ -243,10 +245,11 @@ static Slice build(final Settings settings, final Authentication auth,
                 final String host = cfg.settings()
                     .orElseThrow(() -> new IllegalStateException("Repo settings not found"))
                     .string("host");
-                slice = new DockerSlice(
-                    cfg.path(),
-                    new ProxyDocker(new ClientSlice(SliceFromConfig.HTTP, host))
-                );
+                final Docker proxy = new ProxyDocker(new ClientSlice(SliceFromConfig.HTTP, host));
+                final Docker docker = cfg.storageOpt()
```
Comment (Vatavuk) [inline src/main/java/com/artipie/SliceFromConfig.java:249]: @olegmoz can we add a test for this?

Comment (olegmoz) [reply src/main/java/com/artipie/SliceFromConfig.java:249]: @Vatavuk I doubt that we can test this class in any meaningful way without significant refactoring. We may check that it returns `DockerSlice`, but cannot check what's inside in unit test

Comment (Vatavuk) [reply src/main/java/com/artipie/SliceFromConfig.java:249]: @olegmoz what about an IT test?

Comment (olegmoz) [reply src/main/java/com/artipie/SliceFromConfig.java:249]: @Vatavuk that would be possible, but there no any IT tests in the project yet

File: src/main/java/com/artipie/RepoConfig.java [hunk 2]
```diff
@@ -128,17 +128,32 @@ public URL url() {
...
+            () -> new IllegalStateException("Storage is not configured")
+        );
+    }
+
+    /**
+     * Create storage if configured.
+     *
+     * @return Async storage for repo
+     */
+    public Optional<Storage> storageOpt() {
```
Comment (g4s8) [inline src/main/java/com/artipie/RepoConfig.java:141]: @olegmoz I think we can replace `storage()` with this implementation and just call `storage().orElseThrow()` to get storage where it's required

Comment (olegmoz) [reply src/main/java/com/artipie/RepoConfig.java:141]: @g4s8 I think I did just that, `RepoConfig.storage()` method looks like that now:
```
public Storage storage() {
        return this.storageOpt().orElseThrow(
            () -> new IllegalStateException("Storage is not configured")
        );
    }
```

**Comment by 0crat** [comment]
Job was finished in 6 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @vatavuk[/z](https://www.0crat.com/u/vatavuk), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by Vatavuk** [comment]
@sereshqua ping

**Comment by Vatavuk** [comment]
@g4s8 can you please add quality review here since @sereshqua is not responding

**Comment by g4s8** [comment]
@0crat quality acceptable

**Pull request merged**